test/unit/publishProjectWebViewController.test.ts

```diff
@@ -870,4 +880,233 @@ suite("PublishProjectWebViewController Tests", () => {
+        controller.state.formState.publishTarget = PublishTarget.LocalContainer;
+        controller.state.formState.databaseName = "TestDB";
+        controller.state.formState.containerPort = "1433";
+        controller.state.formState.containerAdminPassword = "MyP@ssw0rd123";
+        controller.state.formState.containerImageTag = "2022-latest";
+        controller.state.formState.acceptContainerLicense = true;
+
+        // Mock prerequisite checks to succeed
+        sandbox
+            .stub(controller, "runDockerPrerequisiteChecks" as keyof typeof controller)
+            .resolves({
+                success: true,
+            });
+
+        // Mock container configuration
+        sandbox
+            .stub(controller, "prepareContainerConfiguration" as keyof typeof controller)
+            .resolves({
+                containerName: "test-container-abc",
+                port: 1433,
+            });
+
+        // Mock container creation to succeed
+        sandbox.stub(controller, "createDockerContainer" as keyof typeof controller).resolves({
+            success: true,
+            connectionUri: "mssql://publish-container-test-container-abc",
+        });
+
+        // Mock build and publish
+        sandbox
+            .stub(controller, "buildProject" as keyof typeof controller)
+            .resolves("/path/to/project.dacpac");
+        sandbox.stub(controller, "publishToDatabase" as keyof typeof controller).resolves();
+
+        // Mock panel dispose
+        const panelDisposeSpy = sandbox.stub();
+        Object.defineProperty(controller, "panel", {
+            value: { dispose: panelDisposeSpy },
+            writable: true,
+            configurable: true,
+        });
+
+        const reducerHandlers = controller["_reducerHandlers"] as Map<string, Function>;
+        const publishNow = reducerHandlers.get("publishNow");
+
+        // Execute publish
+        const newState = await publishNow(controller.state, {});
+
+        // Verify prerequisite checks were run
+        expect(controller["runDockerPrerequisiteChecks"]).to.have.been.calledOnce;
+
+        // Verify container configuration was prepared
+        expect(controller["prepareContainerConfiguration"]).to.have.been.calledOnce;
+
+        // Verify container was created
+        expect(controller["createDockerContainer"]).to.have.been.calledWith(
+            "test-container-abc",
+            1433,
+            sinon.match.any,
+        );
+
+        // Verify build was triggered
+        expect(controller["buildProject"]).to.have.been.calledOnce;
+
+        // Verify publish was triggered
+        expect(controller["publishToDatabase"]).to.have.been.calledOnce;
+
+        // Verify panel was closed at the end
+        expect(panelDisposeSpy).to.have.been.calledOnce;
+
+        // Verify final state
+        expect(newState.inProgress).to.be.false;
+    });
+
+    test("publishNow with LocalContainer handles error scenarios", async () => {
```

Comment:
This seems like it'll be annoying to debug if something goes wrong.  Can you break it up into a different test for each scenario, build the stubs manually for each one, then use a shared helper for doing the error validation?